pkg/loader/bench_test.go

```diff
@@ -176,14 +176,15 @@ func (r *runner) close() {
 func createTable(db *sql.DB, loader Loader) error {
 	sql := "create table test1(id int primary key, a1 int)"
 	// sql = "create table test1(id int, a1 int, UNIQUE KEY `id` (`id`))"
-	loader.Input() <- NewDDLTxn("test", "test1", sql)
+	//loader.Input() <- NewDDLTxn("test", "test1", sql)
+	loader.Publish(NewDDLTxn("test", "test1", sql))
```

Comment:
Could you please paste the output of the `loader/bench_test.go` before and after the change?

Reply:
Ok, I will add more unit-test about disruptors.

Reply:
Cool, correctness comes before performance.

Reply:
Here is the result after run bench_test.go:

|                        | Before                                    | After                                     |   |   |
|-----------------------:|-------------------------------------------|-------------------------------------------|---|---|
| BenchmarkInsertMerge   | 1282 903629 ns/op 2081 B/op 47 allocs/op  | 1488 743148 ns/op 2025 B/op 46 allocs/op  |   |   |
| BenchmarkInsertNoMerge | 1096 1237625 ns/op 2086 B/op 47 allocs/op | 1483 932181 ns/op 1997 B/op 46 allocs/op  |   |   |
| BenchmarkUpdateMerge   | 682 1811123 ns/op 4232 B/op 100 allocs/op | 561 1914113 ns/op 4251 B/op 100 allocs/op |   |   |
| BenchmarkUpdateNoMerge | 546 1905920 ns/op 4232 B/op 100 allocs/op | 684 1674532 ns/op 4131 B/op 99 allocs/op  |   |   |
| BenchmarkDeleteMerge   | 658 1851527 ns/op 3351 B/op 81 allocs/op  | 618 1870392 ns/op 3307 B/op 81 allocs/op  |   |   |
| BenchmarkDeleteNoMerge | 640 1947590 ns/op 3310 B/op 81 allocs/op  | 596 1862614 ns/op 3301 B/op 81 allocs/op  |   |   |

Reply:
Original format when I run like this:

```
go test -benchmem -run=XXX -bench=.
[2019/12/09 18:13:23.593 +07:00] [INFO] [load.go:360] ["exec ddl success"] [sql="drop table if exists test1"]
[2019/12/09 18:13:23.630 +07:00] [INFO] [load.go:360] ["exec ddl success"] [sql="create table test1(id int primary key, a1 int)"]
[2019/12/09 18:13:23.630 +07:00] [INFO] [load.go:264] ["refresh table info"] [schema=test] [table=test1]
goos: darwin
goarch: amd64
pkg: github.com/pingcap/tidb-binlog/pkg/loader
BenchmarkDeleteMerge-4     	[2019/12/09 18:13:23.678 +07:00] [INFO] [load.go:360] ["exec ddl success"] [sql="drop table if exists test1"]
[2019/12/09 18:13:23.698 +07:00] [INFO] [load.go:360] ["exec ddl success"] [sql="create table test1(id int primary key, a1 int)"]
[2019/12/09 18:13:23.698 +07:00] [INFO] [load.go:264] ["refresh table info"] [schema=test] [table=test1]
[2019/12/09 18:13:23.896 +07:00] [INFO] [load.go:360] ["exec ddl success"] [sql="drop table if exists test1"]
[2019/12/09 18:13:23.918 +07:00] [INFO] [load.go:360] ["exec ddl success"] [sql="create table test1(id int primary key, a1 int)"]
[2019/12/09 18:13:23.918 +07:00] [INFO] [load.go:264] ["refresh table info"] [schema=test] [table=test1]
[2019/12/09 18:13:24.145 +07:00] [INFO] [load.go:360] ["exec ddl success"] [sql="drop table if exists test1"]
[2019/12/09 18:13:24.168 +07:00] [INFO] [load.go:360] ["exec ddl success"] [sql="create table test1(id int primary key, a1 int)"]
[2019/12/09 18:13:24.168 +07:00] [INFO] [load.go:264] ["refresh table info"] [schema=test] [table=test1]
[2019/12/09 18:13:24.781 +07:00] [INFO] [load.go:360] ["exec ddl success"] [sql="drop table if exists test1"]
[2019/12/09 18:13:24.805 +07:00] [INFO] [load.go:360] ["exec ddl success"] [sql="create table test1(id int primary key, a1 int)"]
[2019/12/09 18:13:24.805 +07:00] [INFO] [load.go:264] ["refresh table info"] [schema=test] [table=test1]
     618	   1870392 ns/op	    3307 B/op	      81 allocs/op
[2019/12/09 18:13:25.948 +07:00] [INFO] [load.go:360] ["exec ddl success"] [sql="drop table if exists test1"]
[2019/12/09 18:13:25.972 +07:00] [INFO] [load.go:360] ["exec ddl success"] [sql="create table test1(id int primary key, a1 int)"]
[2019/12/09 18:13:25.972 +07:00] [INFO] [load.go:264] ["refresh table info"] [schema=test] [table=test1]
BenchmarkDeleteNoMerge-4   	[2019/12/09 18:13:26.031 +07:00] [INFO] [load.go:360] ["exec ddl success"] [sql="drop table if exists test1"]
[2019/12/09 18:13:26.053 +07:00] [INFO] [load.go:360] ["exec ddl success"] [sql="create table test1(id int primary key, a1 int)"]
[2019/12/09 18:13:26.053 +07:00] [INFO] [load.go:264] ["refresh table info"] [schema=test] [table=test1]
[2019/12/09 18:13:26.187 +07:00] [INFO] [load.go:360] ["exec ddl success"] [sql="drop table if exists test1"]
[2019/12/09 18:13:26.204 +07:00] [INFO] [load.go:360] ["exec ddl success"] [sql="create table test1(id int primary key, a1 int)"]
[2019/12/09 18:13:26.204 +07:00] [INFO] [load.go:264] ["refresh table info"] [schema=test] [table=test1]
[2019/12/09 18:13:26.478 +07:00] [INFO] [load.go:360] ["exec ddl success"] [sql="drop table if exists test1"]
[2019/12/09 18:13:26.500 +07:00] [INFO] [load.go:360] ["exec ddl success"] [sql="create table test1(id int primary key, a1 int)"]
[2019/12/09 18:13:26.500 +07:00] [INFO] [load.go:264] ["refresh table info"] [schema=test] [table=test1]
[2019/12/09 18:13:27.323 +07:00] [INFO] [load.go:360] ["exec ddl success"] [sql="drop table if exists test1"]
[2019/12/09 18:13:27.350 +07:00] [INFO] [load.go:360] ["exec ddl success"] [sql="create table test1(id int primary key, a1 int)"]
[2019/12/09 18:13:27.350 +07:00] [INFO] [load.go:264] ["refresh table info"] [schema=test] [table=test1]
     596	   1862614 ns/op	    3301 B/op	      81 allocs/op
```